openpype/hosts/maya/api/plugin.py

```diff
@@ -143,15 +144,52 @@ def load(
         assert os.path.exists(self.fname), "%s does not exist." % self.fname
 
         asset = context['asset']
+        subset = context['subset']
+        settings = get_project_settings(context['project']['name'])
+        custom_naming = settings['maya']['load']['reference_loader']
         loaded_containers = []
 
+        if not custom_naming['namespace']:
+            raise LoadError("No namespace specified in "
+                            "Maya ReferenceLoader settings")
+        elif not custom_naming['group_name']:
+            raise LoadError("No group name specified in "
+                            "Maya ReferenceLoader settings")
+
+        custom_namespace = custom_naming['namespace'].format(
+            asset_name=asset['name'],
+            asset_type=asset['type'],
+            subset=subset['name'],
+            family=(
+                subset['data'].get('family') or
+                subset['data']['families'][0]
+            )
+        )
+
+        custom_group_name = custom_naming['group_name'].format(
+            asset_name=asset['name'],
+            asset_type=asset['type'],
+            subset=subset['name'],
+            family=(
+                subset['data'].get('family') or
+                subset['data']['families'][0]
+            )
+        )
+
         count = options.get("count") or 1
+
         for c in range(0, count):
-            namespace = namespace or lib.unique_namespace(
-                "{}_{}_".format(asset["name"], context["subset"]["name"]),
-                prefix="_" if asset["name"][0].isdigit() else "",
-                suffix="_",
+            namespace = lib.unique_namespace(
+                custom_namespace,
+                prefix="_" if custom_namespace[0].isdigit() else "",
+                suffix=""
```

Comment:
With the `"_"` not being set in the suffix does that mean that after this PR the naming convention will be slightly different? I'm personally fine with it being removed but it does mean that it'll change the behavior. So I just wanted to highlight that.

Reply:
This is actually a good point. For reference I get `ball_rigMain01:_GRP` with this PR, whereas it was `ball_rigMain_01_:_GRP` before.
I wonder whether we should add another token called `amount` or `identifier` (dont like either of those names), so studios can control where the number gets inserted?
We could get a completely backwards compatible code plus added control.

Reply:
Can't we just do something like this ?
 ```
if custom_namespace[-1] != "_":
    custom_namespace += "_"
```
Let the user chose where the number gets inserterd isn't a bit risky ?

Reply:
> Let the user chose where the number gets inserterd isn't a bit risky ?

If we provide a default value for the namespace that includes the number and warn in the documentation about removing it, I think that should be enough to reduce the risk.

Reply:
This should be pretty decent though:

```python
import re
from openpype.hosts.maya.api.lib import unique_namespace


def get_custom_namespace(custom_namespace):
    """Return unique namespace.

    The input namespace can contain a single group
    of '#' number tokens to indicate where the namespace's
    unique index should go.

    Warning: Note that a namespace will always be
    prefixed with a _ if it starts with a digit

    Examples:
    >>> get_custom_namespace("myspace_##_")
    # myspace_01_
    >>> get_custom_namespace("##_myspace")
    # _01_myspace
    >>> get_custom_namespace("myspace##")
    # myspace01

    """

    if not custom_namespace:
        raise ValueError("Empty custom namespace not supported")

    split = re.split("([#]+)", custom_namespace, 1)
    print(split)
    
    if len(split) == 3:
        # Custom namespace contains # for unique index position
        base, padding, suffix = split
        index_format = "%0{}d".format(len(padding))
    else:
        base = split[0]
        index_format = "%02d" # default behavior
        suffix = ""
        
    return unique_namespace(
        base, 
        format=index_format, 
        prefix="_" if not base or base[0].isdigit() else "",
        suffix=suffix
    )
 
# Testing
for test_custom_namespace in ["{asset_name}_{subset}_##", "{derp}_##_hello", "##hello", "hello"]:
    print(get_custom_namespace(test_custom_namespace))
```

Reply:
Works for me. @friquette how does that sound?

Reply:
Yeah that's sounds great, I will try it :)